### PR TITLE
Remove implicit approval chilling upon slash.

### PR DESCRIPTION
**Pull request by kianenigma**
related to https://github.com/paritytech/substrate/issues/6835 and https://github.com/paritytech/polkadot/pull/5974 and https://github.com/paritytech/polkadot/pull/5978

## (NEW) Context

This was discussed extensively with @burdges and other w3f researchers and the conclusion was that in the short term: 

- the chilling that we do now is too restrictive.
- chilling upon slash enactment (what this PR initially wanted to do) is quite useless.

the ideal behavior we want is for nominators to be immediately chilled, perhaps based on some preference that they set, and then re-activated if the slash is canceled. 

But, for now, and in the short term, we just remove the slashing behavior. 

## (OUTDATED) Context 

The objective is to achieve this statement of the [slashing spec](https://research.web3.foundation/en/latest/polkadot/slashing/npos.html): 

> We first post a slashing transaction to the chain, which drops the offending validator  from the active validator list by invalidating their controller key, or maybe just their session keys. In consequence, all nodes ignore  for the remainder of the era. It invalidates any future blocks that do not ignore  too. **We also remove all nomination approval votes by any nominator for , even those who currently allocate  zero stake.**

Historically, @rphmeier implemented this by reading the slashing spans upon the staking election. Over the years, the staking system has grown and the number of nominators has grown a lot, ergo the election process has become more weight/memory intense. 

> I don't see any reason for to think that the original approach had a particular advantage, but I might be missing something. 

Moreover, the `SlashingSpans` of a staker are only removed ONLY when the staker is fully removed. This means the size of this storage item is potentially very large (see below for ) and it is being read every era upon staking election.

This changes the logic such that this chilling happens upon slash, and it is explicit, meaning that we actually remove the nomination. We still record `SlashingSpans::last_nonzoro_slash` and `Nominations::submitted_in`, which are no longer used onchain, but a UI can still use it to speculate if/when you need to re-send your nominations to a slashed validator (cc @jacogr @rossbulat). 

This will basically make the slashing and npos election part decoupled, and reduces our worries about memory issues when creating the snapshot.  

Regarding https://github.com/paritytech/substrate/issues/6835, now it is pretty easy to move this chilling from `compute_slash` (when slash is reported) to `do_slash` (when slash is applied), so that validators who get slashed and are alter forgiven via governance don't lose nominations. 

## Review notes

The diff is bit noisy because I fixed some comments according to https://github.com/paritytech/substrate/discussions/12404 and a few other random things. The main diff is only in `slashing.rs` and `pallet/impls.rs`

I also added an event for when the slash is reported. Might help UIs. cc @wpank 

## Current slashing spans 

```
// polkadot: 
Len = 455 / encoded_size = 22537
// kusama
Len = 2477 / encoded_size = 124903
```

As you see, we also have a problem of storing too much slashing spans, which I will formulate in a separate issue. 

#12409 Would come very handy here. 

polkadot companion: https://github.com/paritytech/polkadot/pull/6424

### CODE REVIEW

**Comment by burdges** [comment]
We should likely remove the slashing spans after the unbonding period, perhaps by making stakers remove them whenever they touch their account, like unbonding, etc.  We could make their controller key similarly remove the old spans.

We (Rob & I) designed the slashing spans system to make slashing more fair.  We never implemented the whole scheme  We threw this out the window for 100% soundness slashes, if not before then once I noticed that gambler's ruin gives us an actual security proof for the whole system.  We also now know how to make it much harder to mistakenly triggering these non-soundness slashes.  We could therefore discuss more radical simplifications if anyone dares.  ;)

**Comment by kianenigma** [comment]
@ordian @rphmeier have either of you had the time to look at this?

File: frame/staking/src/slashing.rs [hunk 1]
```diff
@@ -411,12 +409,21 @@ fn slash_nominators<T: Config>(
...
 				spans.end_span(params.now);
 			}
 		}
 
+		// in any case, remove the nomination.
+		<Pallet<T> as Store>::Nominators::mutate(
+			&nominator.who,
+			|maybe_nomination: &mut Option<crate::Nominations<T>>| {
+				if let Some(nomination) = maybe_nomination.as_mut() {
+					nomination.targets.retain(|t| t != params.stash)
```
Comment (kianenigma) [inline frame/staking/src/slashing.rs:422]: maybe we should only chill if the slash percentage is more than a threshold, or generally a condition is met.

Comment (kianenigma) [reply frame/staking/src/slashing.rs:422]: generally, we want to, perhaps, not chill nominators, if the slash is zero (offline-ness)

**Review by ordian — COMMENTED** [review]
Generally I'm in favor of this change. However, before approving, I'd like to get clarification on two things:
1. Why was slashing [designed](https://github.com/paritytech/substrate/issues/6835#issuecomment-670108019) so that validators lose nominators even if the slash is reversed?
> We intentionally error in favor of slashed nominators needing to redo their nominations, not just of the slashed validator, but all their nominations. 
2. On zero slashes, validators should not lose nominators IMHO.

**Comment by kianenigma** [comment]
The decision with @burdges and @ordian and robertk was to remove this chilling for the time being.

File: frame/staking/src/pallet/impls.rs [hunk 2]
```diff
@@ -714,18 +714,12 @@ impl<T: Config> Pallet<T> {
...
-						.get(stash)
-						.map_or(true, |spans| submitted_in >= spans.last_nonzero_slash())
-				});
-				if !targets.len().is_zero() {
+			if let Some(Nominations { targets, .. }) = <Nominators<T>>::get(&voter) {
+				if !targets.is_empty() {
 					all_voters.push((voter.clone(), weight_of(&voter), targets));
 					nominators_taken.saturating_inc();
+				} else {
+					// A nominator's nominations might become empty due to slashing.
```
Comment (ordian) [inline frame/staking/src/pallet/impls.rs:722]: Is it true in the current version of the PR?

File: utils/frame/remote-externalities/src/lib.rs [hunk 3]
```diff
@@ -930,7 +935,7 @@ mod remote_tests {
 			))
 			.build()
 			.await
-			.expect(REMOTE_INACCESSIBLE)
+			.unwrap()
```
Comment (ordian) [inline utils/frame/remote-externalities/src/lib.rs:938]: this is an unrelated change, right?

**Review by ordian — APPROVED** [review]
IIUC, not chilling nominators at all on slashes is be a big user-facing change that needs to be communicated.

I'm relatively new to staking pallet, so approving mostly aspiration assuming implementation matches it. Adding an event for slashing also looks good.

CI complains about missing Cargo.lock changes.

File: frame/staking/src/pallet/impls.rs (hunk 2 above)
Comment (kianenigma) [reply frame/staking/src/pallet/impls.rs:722]: yeah not really anymore, but safe to keep it for documentation purposes.

**Comment by kianenigma** [comment]
> IIUC, not chilling nominators at all on slashes is be a big user-facing change that needs to be communicated.

Totally agree, I will post about this in the polkadot forum once it is finalized.

File: utils/frame/remote-externalities/src/lib.rs (hunk 3 above)
Comment (kianenigma) [reply utils/frame/remote-externalities/src/lib.rs:938]: yup.

File: frame/staking/src/tests.rs [hunk 4]
```diff
@@ -2853,6 +2855,9 @@ fn deferred_slashes_are_deferred() {
 			&[Perbill::from_percent(10)],
 		);
 
+		// nominations are removed regardless of the deferring.
```
Comment (eskimor) [inline frame/staking/src/tests.rs:2858]: What does nominations are removed mean? Isn't that chilling?

File: frame/staking/src/tests.rs [hunk 5]
```diff
@@ -2866,24 +2871,23 @@ fn deferred_slashes_are_deferred() {
...
 
-		assert_eq!(
-			staking_events_since_last_call(),
-			vec![
-				Event::StakersElected,
-				Event::EraPaid { era_index: 3, validator_payout: 11075, remainder: 33225 },
+		assert!(matches!(
+			staking_events_since_last_call().as_slice(),
+			&[
+				Event::Chilled { stash: 11 },
```
Comment (eskimor) [inline frame/staking/src/tests.rs:2884]: I am a bit confused. Shouldn't the validator have been chilled before this change already?

File: frame/staking/src/tests.rs [hunk 6]
```diff
@@ -3213,18 +3219,11 @@ fn slash_kicks_validators_not_nominators_and_disables_nominator_for_kicked_valid
...
+		// check that validator was chilled.
+		assert!(<Staking as Store>::Validators::iter().all(|(stash, _)| stash != 11));
 
-		let nominations = <Staking as Store>::Nominators::get(&101).unwrap();
-
-		// and make sure that the vote will be ignored even if the validator
-		// re-registers.
-		let last_slash = <Staking as Store>::SlashingSpans::get(&11).unwrap().last_nonzero_slash();
-		assert!(nominations.submitted_in < last_slash);
+		// and make sure the nominator's approval vote is dropped.
```
Comment (eskimor) [inline frame/staking/src/tests.rs:3225]: What is an approval vote here?

File: frame/staking/src/tests.rs [hunk 7]
```diff
@@ -3256,12 +3255,18 @@ fn non_slashable_offence_doesnt_disable_validator() {
...
+		assert_eq!(Staking::nominators(101).unwrap().targets, vec![11, 21]);
+
 		// offence that slashes 25% of the bond
 		on_offence_now(
 			&[OffenceDetails { offender: (21, exposure_21.clone()), reporters: vec![] }],
 			&[Perbill::from_percent(25)],
 		);
 
+		// it DOES affect the nominator.
+		assert_eq!(Staking::nominators(101).unwrap().targets, vec![11]);
```
Comment (eskimor) [inline frame/staking/src/tests.rs:3268]: So we still chill nominations? :confused:

**Review by eskimor — COMMENTED** [review]
After a few passes and reading up in source code things start to make sense a bit, but I am still confused. Left a few comments, illustrating my confusion.

File: frame/staking/src/pallet/impls.rs [hunk 8]
```diff
@@ -719,7 +720,6 @@ impl<T: Config> Pallet<T> {
```
Comment (ordian) [inline frame/staking/src/pallet/impls.rs:720]: The comments at https://github.com/paritytech/substrate/blob/32abede3de9d86461a045e2f247f0c45edc2826a/frame/staking/src/pallet/impls.rs#L709 are now outdated

File: frame/staking/src/benchmarking.rs [hunk 9]
```diff
@@ -792,12 +792,10 @@ benchmarks! {
...
 
 	get_npos_voters {
-		// number of validator intention.
-		let v in (MaxValidators::<T>::get() / 2) .. MaxValidators::<T>::get();
-		// number of nominator intention.
-		let n in (MaxNominators::<T>::get() / 2) .. MaxNominators::<T>::get();
-		// total number of slashing spans. Assigned to validators randomly.
-		let s in 1 .. 20;
+		// number of validator intention. we will iterate all of them.
+		let v in (MaxValidators::<T>::get() / 4) .. MaxValidators::<T>::get();
```
[the remark this answers is not in the document]
Comment (ruseinov) [reply frame/staking/src/benchmarking.rs:796]: Kinda puzzling to me too, I think this is set to a 100, so this iterates from 25 to a 100. Why this magic number?

File: frame/staking/src/benchmarking.rs [hunk 10]
```diff
@@ -792,12 +792,10 @@ benchmarks! {
...
+		// number of nominator intention. we will iterate all of them.
+		let n in (MaxNominators::<T>::get() / 4) .. MaxNominators::<T>::get();
```
Comment (ruseinov) [inline frame/staking/src/benchmarking.rs:798]: Same here, not sure why this denominator? I can see that this lines up with MaxValidators as they are set to exactly same number for benches (100), but the general idea is not clear.

File: frame/staking/src/pallet/impls.rs [hunk 11]
```diff
@@ -737,18 +737,12 @@ impl<T: Config> Pallet<T> {
...
-						.get(stash)
-						.map_or(true, |spans| submitted_in >= spans.last_nonzero_slash())
-				});
-				if !targets.len().is_zero() {
+			if let Some(Nominations { targets, .. }) = <Nominators<T>>::get(&voter) {
+				if !targets.is_empty() {
 					all_voters.push((voter.clone(), weight_of(&voter), targets));
 					nominators_taken.saturating_inc();
+				} else {
+					// Technically should never happen, but not much we can do about it.
```
Comment (ruseinov) [inline frame/staking/src/pallet/impls.rs:745]: Should there be a debug statement at least? Otherwise why bother with an `else` branch?

File: frame/staking/src/slashing.rs [hunk 12]
```diff
@@ -239,9 +239,9 @@ pub(crate) fn compute_slash<T: Config>(
 		return None
 	}
 
-	let (prior_slash_p, _era_slash) =
+	let prior_slash_p =
 		<Pallet<T> as Store>::ValidatorSlashInEra::get(&params.slash_era, params.stash)
-			.unwrap_or((Perbill::zero(), Zero::zero()));
+			.map_or(Zero::zero(), |(p, _)| p);
```
Comment (ruseinov) [inline frame/staking/src/slashing.rs:244]: maybe best to actually have a full name for this, `prior_slash_p = p` is kinda cryptic.

File: frame/staking/src/pallet/impls.rs (hunk 11 above)
Comment (kianenigma) [reply frame/staking/src/pallet/impls.rs:745]: Just for readability.

Comment (ruseinov) [reply frame/staking/src/pallet/impls.rs:745]: I'd prefer a simple comment someplace to if/else that does nothing. But I would also describe this as personal preference. Empty if/else blocks are sort of an anti-pattern in any programming language, though I must say that breaking the rules isn't always a bad choice.

**Review by gpestana — CHANGES_REQUESTED** [review]
Added a few comments in the tests. If i'm understanding the flow correctly there are still a few assertions that expect the nominators to be chilled right after the offence.

I'm slightly concerned that validators with applied slashes (not reverted) will be able to start validating with the same supports when the chilling period ends. Perhaps we could drop the supports for slashed validators in `fn apply_slash`? The list of nominators that supported the validator are passed in the `struct UnappliedSlash`, so we could mutate the supports of the slashed nominators there.

File: frame/staking/src/tests.rs [hunk 13]
```diff
@@ -2853,6 +2855,9 @@ fn deferred_slashes_are_deferred() {
 			&[Perbill::from_percent(10)],
 		);
 
+		// nominations are removed regardless of the deferring.
+		assert_eq!(Staking::nominators(101).unwrap().targets, vec![21]);
```
Comment (Ank4n) [inline frame/staking/src/tests.rs:2859]: I expected `11` to not be removed from the targets list?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2143474

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2143473

File: frame/staking/src/tests.rs (hunk 7 above)
Comment (kianenigma) [reply frame/staking/src/tests.rs]: assertions are outdated. Updated now.

File: frame/staking/src/tests.rs (hunk 6 above)
Comment (kianenigma) [reply frame/staking/src/tests.rs]: An "approval" is one item in the list of nominations that a nominator puts forth.

File: frame/staking/src/tests.rs (hunk 5 above)
Comment (kianenigma) [reply frame/staking/src/tests.rs:2884]: Yes, it is because I moved `System::reset_events();` further up.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2143711

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2143712

File: frame/staking/src/benchmarking.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (kianenigma) [reply frame/staking/src/benchmarking.rs:796]: I will revert it back to /2.

File: frame/staking/src/pallet/impls.rs (hunk 8 above)
Comment (kianenigma) [reply frame/staking/src/pallet/impls.rs:720]: thank you.

**Comment by kianenigma** [comment]
> Added a few comments in the tests. If i'm understanding the flow correctly there are still a few assertions that expect the nominators to be chilled right after the offence.
> 
> I'm slightly concerned that validators with applied slashes (not reverted) will be able to start validating with the same supports when the chilling period ends. Perhaps we could drop the supports for slashed validators in `fn apply_slash`? The list of nominators that supported the validator are passed in the `struct UnappliedSlash`, so we could mutate the supports of the slashed nominators there.
> 
> I see a few advantages of removing the supports as the slashed is applied:
> 
> * Bad actors will have their support dropped;
> * Increases even further the punishment for bad actors;
> * Avoids the issue with validators being slashed in a loop, which is a valid concern IMO.
> 
> And the costs/complexity of implementing it in `fn apply_slash` seem low.

This was basically the state of the PR as of a few commits ago. Given the timeliness I would suggest we move forward with the current state, and re-introduce any kind of chilling later, in a more planned manner.

To recall, the reason that we thought chilling upon `apply_slash` is not all that useful is that between a slash happens and when it is applied, the "bad validator" have plenty of time to continue doing bad things. Moreover, if they are actually "bad", all of this is pointless since they will immediately re-nominate. So all in all, chilling nominators for the sake of chain security is a lost cause. They will re-nominate. But, recall that they will get slashed, so their "effectiveness" declines. 

Chilling nominators is mostly to protect nominators, which we are dropping now entirely. Nominators are an active role and should manually react to slashes. The new event I added in this PR should help with that.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2143805

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2143806

**Review by gpestana — APPROVED** [review]

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_staking

**Comment by kianenigma** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ pallet pallet_staking

**Comment by kianenigma** [comment]
/cmd queue -c bench-bot $ pallet pallet_staking

**Comment by kianenigma** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_staking

**Comment by kianenigma** [comment]
bot merge

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/kusama-era-4543-slashing/1410/1

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v0-9-35/1509/1
